src/ipa_pc/mod.rs

```diff
@@ -736,6 +736,155 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
         })
     }
 
+    /// Refers to the "Multi-poly multi-point assertions" section of the "Recursive SNARKs based on Marlin" document
+    fn batch_open_individual_opening_challenges<'a>(
+        ck: &Self::CommitterKey,
+        labeled_polynomials: impl Clone + IntoIterator<Item = &'a LabeledPolynomial<G::ScalarField>>,
+        commitments: impl Clone + IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
+        query_set: &QuerySet<G::ScalarField>,
+        opening_challenges: &dyn Fn(u64) -> G::ScalarField,
+        rands: impl Clone + IntoIterator<Item = &'a Self::Randomness>,
+        rng: Option<&mut dyn RngCore>,
+    ) -> Result<Self::BatchProof, Self::Error>
+    where
+        Self::Randomness: 'a,
+        Self::Commitment: 'a,
+    {
+        let batch_time = start_timer!(|| "Multi poly multi point batching.");
+
+        let mut opening_challenge_counter = 0;
+        
+        // lambda
+        let mut cur_challenge = opening_challenges(opening_challenge_counter);
+        opening_challenge_counter += 1;
+
+        let mut query_to_labels_map = BTreeMap::new();
+
+        let poly_map: BTreeMap<_, _> = labeled_polynomials
+            .clone()
+            .into_iter()
+            .map(|poly| (poly.label(), poly))
+            .collect();
+
+        for (label, (point_label, point)) in query_set.iter() {
+            let labels = query_to_labels_map
+                .entry(point_label)
+                .or_insert((point, BTreeSet::new()));
+            labels.1.insert(label);
+        }
+
+        let mut points = vec![];
+
+        // h(X)
+        let mut batch_polynomial = Polynomial::zero();
+
+        for (_point_label, (point, labels)) in query_to_labels_map.into_iter() {
+
+            for label in labels {
+                let labeled_polynomial =
+                    poly_map.get(label).ok_or(Error::MissingPolynomial {
+                        label: label.to_string(),
+                    })?;
+
+                points.push(point);
+
+                // y_i
+                let evaluated_y = labeled_polynomial.polynomial().evaluate(*point); 
+
+                // (p_i(X) - y_i) / (X - x_i)
+                let polynomial = 
+                    &(labeled_polynomial.polynomial() - &Polynomial::from_coefficients_vec(vec![evaluated_y])) 
+                    / 
+                    &Polynomial::from_coefficients_vec(vec![
+                        (G::ScalarField::zero() - point), 
+                        G::ScalarField::one()
+                    ]);
+                
+                // h(X) = SUM( lambda^i * ((p_i(X) - y_i) / (X - x_i)) )
+                batch_polynomial += (cur_challenge, &polynomial);
+
+                // lambda^i
+                cur_challenge = opening_challenges(opening_challenge_counter);
```

Comment:
This implementation squeezes a fresh challenge instead of taking the power of the first one. That's in alignment of the current Marlin verifier gadget, which fully simulates non-native field arithmetics.
If we plan to implement the dlog verifier gadget with simulated non-native field arithemtics in the first step, then we should keep that approach. Otherwise, and in particular later on when we defer arithmetics, we in fact compute the power of the first challenge lambda (after we converted it from endo rep to scalar).